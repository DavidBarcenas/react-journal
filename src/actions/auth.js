import { types } from "../types/types";

export const startLoginEmailPwd = ( email, pwd ) => {
  return ( dispatch ) => {
    setTimeout(() => {
      dispatch(login(123, 'Daveepro'))
    }, 3500);    
  }
}

export const login = ( uid, displayName ) => ({
  type: types.login,
  payload: {
    uid, 
    displayName
  }
})