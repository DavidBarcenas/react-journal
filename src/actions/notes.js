import { db } from "../firebase/firebase-provider";
import { types } from "../types/types";
import { loadNotes } from "../helpers/loadNotes";
import Swal from "sweetalert2";
import { fileUpload } from "../helpers/fileUpload";

export const startNewNote = () => {
  return async ( dispatch, getState ) => {
    const {auth: { uid }} = getState();
    const newNote = {
      title: '',
      body: '',
      date: new Date().getTime()
    }
    const doc = await db.collection(`${ uid }/journal/notes`).add( newNote )
    dispatch(activeNote( doc.id, newNote ))
  }
}

export const activeNote = ( id, note ) => ({
  type: types.noteActive,
  payload: {
    id, 
    ...note
  }
})

export const setNotes = ( notes ) => ({
  type: types.noteLoad,
  payload: notes
})

export const startLoadingNotes = ( uid ) => {
  return async ( dispatch ) => {
    const notes = await loadNotes( uid )
    dispatch(setNotes( notes ))
  }
}

export const startSaveNote = ( note ) => {
  return async ( dispatch, getState ) => {
    const {auth: { uid }} = getState();

    if( !note.url ) { delete note.url }

    const noteToFS = { ...note }
    delete noteToFS.id;

    await db.doc(`${ uid }/journal/notes/${ note.id }`).update(noteToFS)
    dispatch(refreshNote( note.id, noteToFS ))
    Swal.fire('Saved', note.title, 'success')
  }
}

export const refreshNote = ( id, note ) => ({
  type: types.noteUpdated,
  payload: {
    id, note: { id, ...note }
  }
})

export const startUploading = ( file ) => {
  return async ( dispatch, getState ) => {
    const { active: activeNote } = getState().notes;
    const fileUrl = await fileUpload( file )

    console.log('fileUrl', fileUrl)
  }
}