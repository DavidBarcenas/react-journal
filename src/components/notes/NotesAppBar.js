import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startSaveNote } from '../../actions/notes'

export const NotesAppBar = () => {

  const dispatch = useDispatch()
  const { active } = useSelector(state => state.notes)

  const handleSave = () => {
    dispatch(startSaveNote( active ))
  }

  return (
    <div className="notes__appbar">
      <span className="notes__appbar-date"> 28 de Agosto de 2020 </span>
      <div>
        <button className="btn"> pic </button>
        <button className="btn" onClick={ handleSave }> Save </button>
      </div>
    </div>
  )
}
