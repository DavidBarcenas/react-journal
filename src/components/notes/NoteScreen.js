import React, { useEffect, useRef } from 'react'
import { NotesAppBar } from './NotesAppBar'
import { useDispatch, useSelector } from 'react-redux'
import { useForm } from '../../hooks/useForm'
import { activeNote } from '../../actions/notes'

export const NoteScreen = () => {

  const dispatch = useDispatch()
  const { active:note } = useSelector( state => state.notes)
  const activeId = useRef( note.id )
  const [ formValues, handleInputChange, reset ] = useForm(note)

  useEffect(() => {
    if( note.id !== activeId.current ) {
      reset(note)
      activeId.current = note.id
    }
  }, [ note, reset ])

  useEffect(() => {
    dispatch(activeNote(formValues.id, {...formValues}))
  }, [formValues, dispatch])

  return (
    <div className="notes__main">
      <NotesAppBar />

      <div className="notes__content">
        <form>
          <input type="text" placeholder="Some awesome title" name="title" className="notes__input" value={ formValues.title } onChange={ handleInputChange } />
          <textarea name="body" cols="5" rows="10" placeholder="What happend today" className="notes__textarea" value={ formValues.body } onChange={ handleInputChange } ></textarea>
        </form>

        <div className="notes__image">
          { note.url &&
            <img src={ note.url } alt={ note.title } />
          }
        </div>
      </div>
    </div>
  )
}
