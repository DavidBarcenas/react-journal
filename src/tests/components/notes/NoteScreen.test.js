import '@testing-library/jest-dom';
import { mount } from 'enzyme';
import React from 'react'
import { Provider } from 'react-redux';
import { NoteScreen } from '../../../components/notes/NoteScreen';
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk';
import { activeNote } from '../../../actions/notes';


jest.mock('../../../actions/notes', () => ({
  activeNote: jest.fn(),
}))


const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initState = {
  auth: {
    uid: 'd236d',
    name: 'Davee'
  },
  ui: {
    loading: false,
    msgError: null
  },
  notes: {
    notes: [],
    active: {
      id: 1234,
      title: 'Hello',
      body: 'World',
      date: 0
    }
  }
}

let store = mockStore( initState )
store.dispatch = jest.fn()


const wrapper = mount(
  <Provider store={ store }>
    <NoteScreen />
  </Provider>
)

describe('<NotesScreen /> testing', () => {

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('should render correctly', () => {
    expect(wrapper).toMatchSnapshot()
  })
  
  test('should trigger activeNote action', () => {
    wrapper.find('input[name="title"]').simulate('change', {
      target: {
        value: 'My name is Daveepro',
        name: 'title'
      }
    })
    expect(activeNote).toHaveBeenLastCalledWith(
      1234,
      {
        body: 'World',
        title: 'My name is Daveepro',
        id: 1234,
        date: 0
      }
    )
  })
  
})
