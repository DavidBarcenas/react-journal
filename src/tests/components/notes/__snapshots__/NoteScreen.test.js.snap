// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NotesScreen /> testing should render correctly 1`] = `
<div
  className="notes__main"
>
  <div
    className="notes__appbar"
  >
    <span
      className="notes__appbar-date"
    >
       28 de Agosto de 2020 
    </span>
    <input
      id="fileSelector"
      name="file"
      onChange={[Function]}
      style={
        Object {
          "display": "none",
        }
      }
      type="file"
    />
    <div>
      <button
        className="btn"
        onClick={[Function]}
      >
         Pic 
      </button>
      <button
        className="btn"
        onClick={[Function]}
      >
         Save 
      </button>
    </div>
  </div>
  <div
    className="notes__content"
  >
    <form>
      <input
        className="notes__input"
        name="title"
        onChange={[Function]}
        placeholder="Some awesome title"
        type="text"
        value="Hello"
      />
      <textarea
        className="notes__textarea"
        cols="5"
        name="body"
        onChange={[Function]}
        placeholder="What happend today"
        rows="10"
        value="World"
      />
    </form>
    <div
      className="notes__image"
    />
  </div>
  <button
    className="btn notes__btn-delete"
    onClick={[Function]}
  >
    Delete
  </button>
</div>
`;
